src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_internal.clj

```diff
@@ -86,6 +86,29 @@
    pool :- jruby-schemas/pool-queue-type]
   (.borrowItemWithTimeout pool timeout TimeUnit/MILLISECONDS))
 
+(schema/defn insert-shutdown-poison-pill
+  [pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (ShutdownPoisonPill. pool)))
+
+(schema/defn insert-poison-pill
+  [error :- Throwable
+   pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (PoisonPill. error)))
+
+(schema/defn fill-pool-with-poison-pills!
+  "Fill an empty pool with poison pills. Should always be run on the
+  modify-instance-agent. insert-fn should be a fn that inserts a poison
+  pill of some kind into the pool"
+  [pool-state :- jruby-schemas/PoolState
+   insert-fn :- IFn]
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)]
+    (when (not (= pool-size (.remainingCapacity pool)))
```

Comment:
Hmm, since creating pills will always happen on the modify-instance-agent I'm not too worried about overlapping pill insertions. This should only happen when you're shutting down, and 2 shutdown requests coming in at the same time won't step on each other's toes, since the second request will be waiting in the `.lock` call until the pills have filled the queue.

But regardless, I think it would make more sense conceptually to do what you're saying: have one slot for a pill in the pool, and use it to make decisions. Since the pills are only for shutting down/failures, that seems good to me. Much less ambiguity about how the pills work, and none of this filling the pool with pills nonsense

And yeah, I agree that if there is a pill in the pool no one should be able to get anything back from a borrow besides that pill